services/webhook/deliver.go

```diff
@@ -170,18 +171,32 @@ func Deliver(ctx context.Context, t *webhook_model.HookTask) error {
 		}
 	}()
 
+	// OK We're now ready to attempt to deliver the task - we must double check that it
+	// has not been delivered in the meantime
+	updated, err := webhook_model.MarkTaskDelivered(ctx, t)
```

Comment:
Should we move these lines after line 194?

Reply:
> > how does it get into the queue again to be double delivered?

In concurrent systems all bets are off. 

OK, let's imagine a system that shutsdown with hook tasks 1–6 in the queue. All of these are `is_delivered=false`.

The HookQueue is started as a goroutine and starts handling tasks independently. Lets say it sends 1 & 2 leaving 3–6 in the queue

Then the population script runs. It queries the DB and finds 3–6 aren't delivered. They're returned a tasks to be pushed to the queue.

In the meantime 3 is handled by the queue.

Now we push 3–6 to the queue. 3 is not in the queue so it is added i.e. the queue contains 4,5,6,3.

Thus 3 gets a double delivery.

The situation is even more likely if you imagine a cluster of Gitea's starting up at different times and pushing to a common redis queue.

Reply:
as above